File: modbus_devices/sht30.py
import subprocess

def get_values():
	"""
	Вызывает внешнюю утилиту mbpoll с заданными параметрами,
	извлекает из вывода два числа и возвращает их
	в виде форматированных строк, а также сырые значения.

	Returns:
		tuple: (humidity_str, temperature_str, humidity_raw, temperature_raw)
			   или (None, None, None, None) при ошибке.
	"""
	cmd = ["mbpoll", "-c2", "-1", "-q", "-t4:int16", "192.168.1.40"]

	try:
		result = subprocess.run(cmd, capture_output=True, text=True, check=True)
	except (subprocess.CalledProcessError, FileNotFoundError) as e:
		sys.stderr.write(f"Ошибка получения данных: {e}\n")
		return None, None, None, None

	lines = result.stdout.strip().splitlines()
	if len(lines) < 3:
		sys.stderr.write("Недостаточно строк в выводе утилиты\n")
		return None, None, None, None

	humidity_raw = int(lines[1].split()[1]) / 10
	temperature_raw = int(lines[2].split()[1]) / 10

	humidity_str = f"{humidity_raw} %".ljust(7)
	temperature_str = f"{temperature_raw} c".ljust(7)

	return humidity_str, temperature_str, humidity_raw, temperature_raw

File: sht30.lua
function get_values(ip, modbus_addr)
    -- Изменён тип регистров: -t4:int16
    local cmd = "mbpoll -c2 -1 -q -t4:int16 -a" .. modbus_addr .. " " .. ip

    local p = io.popen(cmd, 'r')
    if not p then
        io.stderr:write("Ошибка запуска утилиты mbpoll\n")
        return nil, nil, nil, nil
    end

    local output = p:read("*a")
    local ok, err = p:close()
    if not ok then
        io.stderr:write(string.format("Ошибка получения данных: %s\n", err))
        return nil, nil, nil, nil
    end

    output = output:gsub("[\r\n]+$", "")
    local lines = {}
    for line in output:gmatch("[^\r\n]+") do
        table.insert(lines, line)
    end

    if #lines < 3 then
        io.stderr:write("Недостаточно строк в выводе утилиты\n")
        return nil, nil, nil, nil
    end

    local function get_second_field(line)
        local parts = {}
        for part in line:gmatch("%S+") do
            table.insert(parts, part)
        end
        if #parts >= 2 then
            return tonumber(parts[2])
        end
        return nil
    end

    -- Порядок извлечения изменён: humidity из lines[2] (вторая строка), temperature из lines[3] (третья)
    local humidity_raw_num = get_second_field(lines[2])
    local temperature_raw_num = get_second_field(lines[3])

    if not humidity_raw_num or not temperature_raw_num then
        io.stderr:write("Невозможно извлечь числа из вывода\n")
        return nil, nil, nil, nil
    end

    local humidity_raw = humidity_raw_num / 10
    local temperature_raw = temperature_raw_num / 10

    -- Выравнивание до 6 символов
    local humidity_str = string.format("%-6s", string.format("%.1f %%", humidity_raw))
    local temperature_str = string.format("%-6s", string.format("%.1f c", temperature_raw))

    return humidity_str, temperature_str, humidity_raw, temperature_raw
end
